src/linear/updater_coordinate.cc

```diff
@@ -0,0 +1,133 @@
+/*!
+ * Copyright 2018 by Contributors
+ * \author Rory Mitchell
+ */
+
+#include <xgboost/linear_updater.h>
+#include "coordinate_common.h"
+#include "../common/timer.h"
+
+namespace xgboost {
+namespace linear {
+
+DMLC_REGISTRY_FILE_TAG(updater_coordinate);
+
+// training parameter
+struct CoordinateTrainParam : public dmlc::Parameter<CoordinateTrainParam> {
+  /*! \brief learning_rate */
+  float learning_rate;
+  /*! \brief regularization weight for L2 norm */
+  float reg_lambda;
+  /*! \brief regularization weight for L1 norm */
+  float reg_alpha;
+  std::string coordinate_selection;
+  float maximum_weight;
+  int debug_verbose;
+  // declare parameters
+  DMLC_DECLARE_PARAMETER(CoordinateTrainParam) {
+    DMLC_DECLARE_FIELD(learning_rate)
+        .set_lower_bound(0.0f)
+        .set_default(1.0f)
+        .describe("Learning rate of each update.");
+    DMLC_DECLARE_FIELD(reg_lambda)
+        .set_lower_bound(0.0f)
+        .set_default(0.0f)
+        .describe("L2 regularization on weights.");
+    DMLC_DECLARE_FIELD(reg_alpha)
+        .set_lower_bound(0.0f)
+        .set_default(0.0f)
+        .describe("L1 regularization on weights.");
+    DMLC_DECLARE_FIELD(coordinate_selection)
+        .set_default("cyclic")
+        .describe("Coordinate selection algorithm");
+    DMLC_DECLARE_FIELD(maximum_weight)
+        .set_lower_bound(0.0f)
+        .set_default(1e15f)
+        .describe(
+            "Maximum absolute value of linear model weights. Prevents "
+            "coefficients increasing to infinity in separable logistic "
+            "regression problems.");
+    DMLC_DECLARE_FIELD(debug_verbose)
+        .set_lower_bound(0)
+        .set_default(0)
+        .describe("flag to print out detailed breakdown of runtime");
+    // alias of parameters
+    DMLC_DECLARE_ALIAS(reg_lambda, lambda);
+    DMLC_DECLARE_ALIAS(reg_alpha, alpha);
+  }
+};
+
+/**
+ * \class CoordinateUpdater
+ *
+ * \brief Coordinate descent algorithm that updates one feature per iteration
+ */
+
+class CoordinateUpdater : public LinearUpdater {
+ public:
+  // set training parameter
+  void Init(
+      const std::vector<std::pair<std::string, std::string> > &args) override {
+    param.InitAllowUnknown(args);
+    selector.reset(CoordinateSelector::Create(param.coordinate_selection));
+    monitor.Init("GBLinear ", param.debug_verbose);
```

Comment:
A second "GBLinear " Monitor? Would it make sense to pass GBLinear's monitor to updaters?

Reply:
I forgot to change its name when I copied it. I think it's fine for the updater to have its own monitor as its not really necessary for it to be in the interface.